File: examples/clj/src/com/rpl/agent/basic/streaming_agent.clj
(ns com.rpl.agent.basic.streaming-agent
  "Demonstrates streaming chunk emission from agent nodes.

  Features demonstrated:
  - stream-chunk!: Emit streaming data from nodes
  - agent-stream: Subscribe to streaming data from specific nodes
  - Real-time data flow with incremental results
  - Streaming completion and callbacks"
  (:require
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]))

;;; Agent module demonstrating streaming functionality
(aor/defagentmodule StreamingAgentModule
  [topology]

  (->
   (aor/new-agent topology "StreamingAgent")

    ;; Node that processes data and streams progress
   (aor/node
    "process-data"
    nil
    (fn [agent-node {:keys [data-size chunk-size]}]
      (let [total-chunks (int (Math/ceil (/ data-size chunk-size)))]
        (println
         (format "Processing %d items in chunks of %d" data-size chunk-size))

         ;; Stream progress as we process chunks
        (doseq [chunk-num (range total-chunks)]
          (let [start-idx (* chunk-num chunk-size)
                end-idx (min (+ start-idx chunk-size) data-size)
                items (range start-idx end-idx)
                progress (double (/ (inc chunk-num) total-chunks))]

             ;; Simulate processing time
            (Thread/sleep 100)

             ;; Stream chunk progress
            (aor/stream-chunk! agent-node
                               {:chunk-number chunk-num
                                :items-processed (count items)
                                :progress progress
                                :items items})

            (println (format "Processed chunk %d/%d (%.1f%%)"
                             (inc chunk-num)
                             total-chunks
                             (double (* progress 100))))))

         ;; Return final result
        (aor/result! agent-node
                     {:action "data-processing"
                      :total-items data-size
                      :total-chunks total-chunks
                      :chunk-size chunk-size
                      :completed-at (System/currentTimeMillis)}))))))

(defn -main
  "Run the streaming agent example"
  [& _args]
  (with-open [ipc (rtest/create-ipc)]
    (rtest/launch-module! ipc StreamingAgentModule {:tasks 1 :threads 1})

    (let [manager (aor/agent-manager ipc
                                     (rama/get-module-name
                                      StreamingAgentModule))
          agent (aor/agent-client manager "StreamingAgent")]

      (println "Streaming Agent Example:")
      (println "Processing data with real-time streaming updates...")

      ;; Start async processing
      (let [invoke (aor/agent-initiate agent
                                       {:data-size 50
                                        :chunk-size 10})
            chunks-received (atom [])]

        ;; Subscribe to streaming chunks
        (aor/agent-stream
         agent
         invoke
         "process-data"
         (fn [all-chunks new-chunks reset? complete?]
           (doseq [chunk new-chunks]
             (swap! chunks-received conj chunk)
             (println (format "Received chunk %d: %d items (%.1f%% complete)"
                              (:chunk-number chunk)
                              (:items-processed chunk)
                              (* (:progress chunk) 100))))))

        ;; Wait for completion
        (let [result (aor/agent-result agent invoke)]
          (println "\nFinal result:")
          (println "  Total items processed:" (:total-items result))
          (println "  Total chunks:" (:total-chunks result))
          (println "  Chunk size:" (:chunk-size result))
          (println "  Chunks received via streaming:" (count @chunks-received)))

        (println "\nNotice how:")
        (println "- Streaming provides real-time progress updates")
        (println "- Chunks are received while processing continues")
        (println "- Final result provides summary information")))))

(comment
  (-main))

File: examples/java/basic/src/main/java/com/rpl/agent/basic/StreamingAgent.java
package com.rpl.agent.basic;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentInvoke;
import com.rpl.agentorama.AgentManager;
import com.rpl.agentorama.AgentNode;
import com.rpl.agentorama.AgentsModule;
import com.rpl.agentorama.AgentTopology;
import com.rpl.agentorama.ops.RamaVoidFunction2;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Java example demonstrating streaming chunk emission from agent nodes.
 *
 * <p>Features demonstrated:
 *
 * <ul>
 *   <li>streamChunk: Emit streaming data from nodes
 *   <li>agent.stream: Subscribe to streaming data from specific nodes
 *   <li>Real-time data flow with incremental results
 *   <li>Streaming completion and callbacks
 * </ul>
 *
 * <p>All required classes are defined as nested classes within this single file for simplicity and
 * self-containment.
 */
public class StreamingAgent {

  /** Agent Module demonstrating streaming functionality. */
  public static class StreamingAgentModule extends AgentsModule {

    @Override
    protected void defineAgents(AgentTopology topology) {
      topology.newAgent("StreamingAgent").node("process-data", null, new ProcessDataFunction());
    }
  }

  /** Node function that processes data and streams progress updates. */
  public static class ProcessDataFunction
      implements RamaVoidFunction2<AgentNode, Map<String, Object>> {

    @Override
    public void invoke(AgentNode agentNode, Map<String, Object> request) {
      int dataSize = (Integer) request.get("dataSize");
      int chunkSize = (Integer) request.get("chunkSize");
      int totalChunks = (int) Math.ceil((double) dataSize / chunkSize);

      System.out.printf("Processing %d items in chunks of %d%n", dataSize, chunkSize);

      // Stream progress as we process chunks
      for (int chunkNum = 0; chunkNum < totalChunks; chunkNum++) {
        int startIdx = chunkNum * chunkSize;
        int endIdx = Math.min(startIdx + chunkSize, dataSize);
        List<Integer> items = new ArrayList<>();
        for (int i = startIdx; i < endIdx; i++) {
          items.add(i);
        }
        double progress = (double) (chunkNum + 1) / totalChunks;

        // Simulate processing time
        try {
          Thread.sleep(100);
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
          return;
        }

        // Stream chunk progress
        Map<String, Object> chunkData = new HashMap<>();
        chunkData.put("chunkNumber", chunkNum);
        chunkData.put("itemsProcessed", items.size());
        chunkData.put("progress", progress);
        chunkData.put("items", items);
        agentNode.streamChunk(chunkData);

        System.out.printf(
            "Processed chunk %d/%d (%.1f%%)%n", chunkNum + 1, totalChunks, progress * 100.0);
      }

      // Return final result
      Map<String, Object> result = new HashMap<>();
      result.put("action", "data-processing");
      result.put("totalItems", dataSize);
      result.put("totalChunks", totalChunks);
      result.put("chunkSize", chunkSize);
      result.put("completedAt", System.currentTimeMillis());
      agentNode.result(result);
    }
  }

  public static void main(String[] args) throws Exception {
    System.out.println("Starting Streaming Agent Example...");

    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Launch the agent module
      StreamingAgentModule module = new StreamingAgentModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("StreamingAgent");

      System.out.println("Streaming Agent Example:");
      System.out.println("Processing data with real-time streaming updates...");

      // Start async processing
      Map<String, Object> request = new HashMap<>();
      request.put("dataSize", 50);
      request.put("chunkSize", 10);
      AgentInvoke invoke = agent.initiate(request);
      AtomicInteger chunksReceived = new AtomicInteger(0);

      // Subscribe to streaming chunks
      agent.stream(
          invoke,
          "process-data",
          (allChunks, newChunks, reset, complete) -> {
            for (Object chunkObj : newChunks) {
              @SuppressWarnings("unchecked")
              Map<String, Object> chunk = (Map<String, Object>) chunkObj;
              chunksReceived.incrementAndGet();
              System.out.printf(
                  "Received chunk %d: %d items (%.1f%% complete)%n",
                  (Integer) chunk.get("chunkNumber"),
                  (Integer) chunk.get("itemsProcessed"),
                  (Double) chunk.get("progress") * 100.0);
            }
          });

      // Wait for completion
      @SuppressWarnings("unchecked")
      Map<String, Object> result = (Map<String, Object>) agent.result(invoke);

      System.out.println("\nFinal result:");
      System.out.println("  Total items processed: " + result.get("totalItems"));
      System.out.println("  Total chunks: " + result.get("totalChunks"));
      System.out.println("  Chunk size: " + result.get("chunkSize"));
      System.out.println("  Chunks received via streaming: " + chunksReceived.get());

      System.out.println("\nNotice how:");
      System.out.println("- Streaming provides real-time progress updates");
      System.out.println("- Chunks are received while processing continues");
      System.out.println("- Final result provides summary information");
    }
  }
}
